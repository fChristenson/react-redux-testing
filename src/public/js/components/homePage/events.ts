export const CURRENT_USER = "HOME_PAGE/CURRENT_USER";
