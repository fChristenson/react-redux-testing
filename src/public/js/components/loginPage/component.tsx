import React from "react";

export class LoginPage extends React.Component<any> {
  constructor(props: any) {
    super(props);
  }

  render() {
    return (
      <div>
        <h1>Login</h1>
        <form onSubmit={this.props.onSubmit}>
          <input name="username" />
          <input type="password" name="password" />
          <input type="submit" />
        </form>
        {this.props.showError && <p>Login failed</p>}
      </div>
    );
  }
}
