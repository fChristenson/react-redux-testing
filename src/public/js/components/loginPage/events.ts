export const LOGIN_ERROR = "LOGIN_PAGE/LOGIN_ERROR";
