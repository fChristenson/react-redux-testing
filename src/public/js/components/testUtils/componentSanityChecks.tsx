import React from "react";
import { shallow } from "enzyme";

// The sanity checks, these are checks that verify that we haven't made a mistake and that our tools work
export const componentSanityChecks = (
  label: string,
  Element: React.ReactElement
) =>
  describe(`Sanity checks for ${label}`, () => {
    it("is defined", () => {
      expect(Element).toBeDefined();
    });

    it("renders the element", () => {
      const wrapper = shallow(Element);
      const expected = false;
      const actual = wrapper.isEmptyRender();
      expect(expected).toEqual(actual);
    });

    it("matches the snapshot", () => {
      const wrapper = shallow(Element);
      const expected = wrapper.html();
      expect(expected).toMatchSnapshot();
    });
  });
