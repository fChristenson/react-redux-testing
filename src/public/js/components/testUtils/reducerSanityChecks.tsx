import { action } from "../../../../lib/action";

// The sanity checks, these are checks that verify that we haven't made a mistake and that our tools work
export const reducerSanityChecks = (
  label: string,
  reducer: (state: any, action: any) => any
) =>
  describe(`Sanity checks for ${label}`, () => {
    it("is defined", () => {
      expect(reducer).toBeDefined();
    });

    it("returns the same state as passed in if no action matches", () => {
      const expected = { foo: 1 };
      const actual = reducer({ foo: 1 }, action("UNMATCHED_ACTION", null));
      expect(expected).toEqual(actual);
    });

    it("returns a initial state if none is passed in", () => {
      const actual = reducer(undefined, action("UNMATCHED_ACTION", null));
      expect(actual).toBeDefined();
      expect(typeof actual).toBe("object");
      expect(actual).not.toBe(null);
      expect(actual).not.toBe(undefined);
      expect(Array.isArray(actual)).toBe(false);
    });
  });
