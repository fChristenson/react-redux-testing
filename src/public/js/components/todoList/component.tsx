import React from "react";

export class TodoList extends React.Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = { isTyping: false };
    this.onChange = this.onChange.bind(this);
  }

  componentWillMount() {
    this.props.getTodos();
  }

  onChange(event: any) {
    if (event.target.value.length > 0) {
      this.setState({ isTyping: true });
    } else {
      this.setState({ isTyping: false });
    }
  }

  render() {
    return (
      <div>
        <h2>Todos</h2>
        <ul>
          {this.props.todos.map((todo: string, i: number) => (
            <li key={i}>{todo}</li>
          ))}
        </ul>
        <form onSubmit={this.props.onSubmit}>
          <input onChange={this.onChange} name="todo" />
        </form>
        {this.state.isTyping && <p>You are typing...</p>}
      </div>
    );
  }
}
