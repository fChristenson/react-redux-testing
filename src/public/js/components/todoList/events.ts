export const ADD_TODO = "TODO_LIST/ADD_TODO";
export const SET_TODOS = "TODO_LIST/SET_TODOS";
