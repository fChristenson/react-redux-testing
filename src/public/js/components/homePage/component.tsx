import React from "react";
import { TodoList } from "../todoList";

/**
 * We are not unit testing this component because this is a page component and such a component will over time
 * grow to contain many different subcomponents and features that will quickly become very slow to maintain
 * through unit testing.
 *
 * To test this component you would have to retest each level of the component tree over and over to assert
 * that the expected behavior of a subcomponent is still working in the larger component.
 *
 * If the subcomponents are made up of well tested standard components testing at this level can be
 * focused on the interaction with features.
 *
 * A more flexible approach to this problem is to deal with it through end to end tests that simulate a user
 * interacting with a feature.
 */
export class HomePage extends React.Component<any> {
  constructor(props: any) {
    super(props);
  }

  componentWillMount() {
    this.props.getCurrentUser();
  }

  render() {
    if (!this.props.user) return null;
    return (
      <div>
        <h1>Home</h1>
        <TodoList />
        <a href="/logout">logout</a>
      </div>
    );
  }
}
