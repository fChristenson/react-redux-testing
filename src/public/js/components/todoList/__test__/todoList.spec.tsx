import React from "react";
import { mapDispatchToProps } from "../";
import { TodoList } from "../component";
import { componentSanityChecks } from "../../testUtils/componentSanityChecks";
import { action } from "../../../../../lib/action";
import { ADD_TODO, SET_TODOS } from "../events";
import { shallow } from "enzyme";

const noop = () => {};

describe("TodoList", () => {
  /**
   * Thanks to snapshot testing and a pure visual component we can see that whatever state
   * we create by passing props in will be easy to test.
   */
  componentSanityChecks(
    "TodoList",
    <TodoList todos={["foo"]} getTodos={noop} onSubmit={noop} />
  );
  componentSanityChecks(
    "TodoList - empty",
    <TodoList todos={[]} getTodos={noop} onSubmit={noop} />
  );

  /**
   * Enzyme is great to use when we have an internal state without side effects
   * but it is worth noting that if this component is covered in a end to end test
   * we are likely going to test this behavior twice as it stands to reason that
   * whoever writes either of these tests will check if the typing behavior is correct.
   *
   * I personally prefer to cover tests like these in the end to end test if I can but if you are
   * making a standard component this is a very useful practice.
   */
  it("should display a message when the user is typing", () => {
    const wrapper = shallow(
      <TodoList todos={[]} getTodos={noop} onSubmit={noop} />
    );

    wrapper
      .find('input[name="todo"]')
      .simulate("change", { target: { value: "test input" } });
    wrapper.update();

    const actual = wrapper.find("p").text();
    const expected = "You are typing...";
    expect(expected).toEqual(actual);
  });

  it("should remove the user is typing message when the input is empty", () => {
    const wrapper = shallow(
      <TodoList todos={[]} getTodos={noop} onSubmit={noop} />
    );

    wrapper
      .find('input[name="todo"]')
      .simulate("change", { target: { value: "test input" } });
    wrapper.update();

    wrapper
      .find('input[name="todo"]')
      .simulate("change", { target: { value: "" } });
    wrapper.update();

    const actual = wrapper.find("p").exists();
    const expected = false;
    expect(expected).toEqual(actual);
  });

  it("should submit the todo form", () => {
    const onsubmit = jest.fn();
    const wrapper = shallow(
      <TodoList todos={[]} getTodos={noop} onSubmit={onsubmit} />
    );

    wrapper
      .find('input[name="todo"]')
      .simulate("change", { target: { value: "test input" } });
    wrapper.update();
    wrapper.find("form").simulate("submit");

    expect(onsubmit).toBeCalledTimes(1);
  });

  /**
   * The following tests all have side effects so to capture this logic we want to stub off the
   * problematic network calls and by moving this logic outside of our component it becomes easy
   * to test it without the need to execute the code through our component.
   */
  describe("mapDispatchToProps", () => {
    const mockDispatch = jest.fn();

    beforeEach(() => {
      mockDispatch.mockClear();
    });

    describe("getTodos", () => {
      it("gets the todos from the server", async () => {
        const mockRes = {
          json: jest.fn().mockReturnValue(["foo"])
        };
        const mockFetch = jest.fn().mockReturnValue(Promise.resolve(mockRes));
        /**
         * By passing in fetch we gain the ability to easily test network logic that would be hard
         * to test otherwise.
         *
         * NOTE: This is not a substitute for using end to end tests since we have no guarantees of
         * what the server will do when we are running the actual system, this test is just to assert
         * that if the component gets what it wants from the server the behavior will be correct.
         */
        const { getTodos } = mapDispatchToProps(mockFetch)(mockDispatch);
        await getTodos();
        expect(mockFetch).toBeCalledTimes(1);
        expect(mockFetch).toBeCalledWith("/api/v1/todos");
        expect(mockDispatch).toBeCalledTimes(1);
        expect(mockDispatch).toBeCalledWith(action(SET_TODOS, ["foo"]));
      });
    });

    describe("onSubmit", () => {
      it("sends a todo to the server", async () => {
        const event = {
          preventDefault: jest.fn(),
          target: {
            todo: {
              value: "foo"
            }
          }
        };
        const opt = {
          method: "POST",
          headers: {
            "Content-Type": "application/json"
          },
          body: JSON.stringify({
            todo: "foo"
          })
        };
        const mockFetch = jest.fn().mockReturnValue(Promise.resolve());
        const { onSubmit } = mapDispatchToProps(mockFetch)(mockDispatch);
        await onSubmit(event);
        expect(event.preventDefault).toBeCalledTimes(1);
        expect(mockFetch).toBeCalledTimes(1);
        expect(mockFetch).toBeCalledWith("/api/v1/todos", opt);
        expect(mockDispatch).toBeCalledTimes(1);
        expect(mockDispatch).toBeCalledWith(action(ADD_TODO, "foo"));
      });
    });
  });
});
