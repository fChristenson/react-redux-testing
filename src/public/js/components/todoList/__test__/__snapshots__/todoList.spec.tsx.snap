// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Sanity checks for TodoList - empty matches the snapshot 1`] = `"<div><h2>Todos</h2><ul></ul><form><input name=\\"todo\\"/></form></div>"`;

exports[`TodoList Sanity checks for TodoList matches the snapshot 1`] = `"<div><h2>Todos</h2><ul><li>foo</li></ul><form><input name=\\"todo\\"/></form></div>"`;
